engines/ic3base.h

```diff
@@ -0,0 +1,428 @@
+**   Makai Mann, Ahmed Irfan
+** This file is part of the pono project.
+** Copyright (c) 2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief Abstract base class implementation of IC3 parameterized by
+**        the unit used in frames, pre-image computation, and inductive
+**        and predecessor generalization techniques.
+**
+**        To create a particular IC3 instantiation, you must implement the
+*following:
+**           - implement get_ic3_formula and give it semantics to produce the
+**             corresponding IC3Formula for your flavor of IC3
+**             (assumes solver_'s state is SAT from a failed rel_ind_check)
+**           - implement inductive_generalization
+**           - implement generalize_predecessor
+**           - implement check_ts which just checks if there are any
+**             theories / syntax used in the transition system which
+**             is not supported by this instantiation
+**          [OPTIONAL]
+**           - implement all the ic3_formula_* functions for creating and
+**             manipulating an IC3Formula if the defaults are not right
+**           - implement abstract() and refine() if this is a CEGAR
+**             flavor of IC3
+**
+**/
+#pragma once
+
+#include "engines/prover.h"
+#include "smt-switch/utils.h"
+
+namespace pono {
+
+struct IC3Formula
+{
+  // nullary constructor
+  IC3Formula() {}
+  IC3Formula(const smt::Term & t, const smt::TermVec & c, bool n)
+      : term(t), children(c), disjunction(n)
+  {
+  }
+
+  IC3Formula(const IC3Formula & other)
+      : term(other.term),
+        children(other.children),
+        disjunction(other.disjunction)
+  {
+  }
+
+  virtual ~IC3Formula() {}
+
+  /** Returns true iff this IC3Formula has not been initialized */
+  bool is_null() const { return (term == nullptr); };
+
+  bool is_disjunction() const { return disjunction; };
+
+  smt::Term term;  ///< term representation of this formula
+  smt::TermVec
+      children;  ///< flattened children of either a disjunction or conjunction
+  bool disjunction;  ///< true if currently representing a disjunction
+  // NOTE: treating the disjunction as the primary orientation
+  //       e.g. aligned with what's kept in frames (disjunctions), not proof
+  //       goals (conjunctions), but IC3Formula can represent both
+};
+
+// TODO change back to ProofGoal once refactor is done
+// don't want to clash with name in MBIC3 for now
+struct IC3Goal
+{
+  // based on open-source ic3ia ProofObligation
+  IC3Formula target;
+  size_t idx;
+  // TODO: see if we can make this a unique_ptr
+  //       made it complicated to move from this struct to another place
+  std::shared_ptr<IC3Goal> next;
+
+  IC3Goal(IC3Formula u, size_t i, std::shared_ptr<IC3Goal> n)
```

Comment:
Why not `const std::shared_ptr<IC3Goal> &`? So you can save one copy in the argument.

Reply:
Great catch! That must have been overlooked because it was a unique pointer at one point that got moved instead of copied. Addressed your comment with d3bd2b4959000b9782a5e431d007b4ae0d549a39